spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/ops/DataFlowOps.scala

```diff
@@ -0,0 +1,586 @@
+
+    val size = tensorArray.shapeOf(0)
+    size(0) = 0
+    val lengths = Tensor[Int](tensorArray.size)
+    var i = 0
+    while(i < tensorArray.size) {
+      size(0) += tensorArray.shapeOf(i)(0)
+      lengths.setValue(i + 1, tensorArray.shapeOf(i)(0))
+      i += 1
+    }
+
+    val value = Tensor[D]().resize(size)
+    i = 0
+    var index = 1
+    while(i < tensorArray.size) {
+      val curSize = tensorArray.shapeOf(i)(0)
+      value.narrow(1, index, curSize).copy(tensorArray(i))
+      index += curSize
+      i += 1
+    }
+
+    output = T(value, lengths)
+    output
+  }
+}
+
+/**
+ * Split the data from the input value into TensorArray elements. It is a 'reverse' operation of
+ * concat. It accepts:
+ *   handle: The handle to a TensorArray.
+ *   value: The concatenated tensor to write to the TensorArray.
+ *   lengths: The vector of lengths, how to split the rows of value into the TensorArray.
+ *
+ * It return a control flow object.
+ *
+ * Assuming that `lengths` takes on values
+ *   (n0, n1, ..., n(T-1))
+ * and that value has shape
+ *   (n0 + n1 + ... + n(T-1) x d0 x d1 x ...),
+ * this splits values into a TensorArray with T tensors.
+ * TensorArray index t will be the subtensor of values with starting position
+ *   `(n0 + n1 + ... + n(t-1), 0, 0, ...)```
+ * and having size
+ *   nt x d0 x d1 x ...
+ *
+ * @tparam T Model parameter numeric type.
+ * @tparam D Element numeric type in the tensor array.
+ */
+private[bigdl] class TensorArraySplit[T: ClassTag, D: ClassTag]()(
+  implicit ev: TensorNumeric[T], ev2: TensorNumeric[D])
+  extends Operation[Table, Tensor[Float], T] {
+
+  output = TensorArray.FlowOut
+
+  override def updateOutput(input: Table): Tensor[Float] = {
+    val handle = input[Tensor[String]](1)
+    val value = input[Tensor[D]](2)
+    val lengths = input[Tensor[Int]](3)
+    require(handle.isScalar, "Handle of a TensorArray must be a scalar")
+    require(lengths.nDimension() == 1, "lengths must be a vector")
+
+    val tensorArray = TensorArray[D](handle.value())
+
+    var i = 1
+    var index = 1
+    while(i <= lengths.size(1)) {
+      tensorArray(i - 1) = value.narrow(1, index, lengths.valueAt(i))
+      index += lengths.valueAt(i)
+      i += 1
+    }
+
+    output
+  }
+}
+
+/**
+ * Get the current size of the TensorArray.
+ *
+ * @tparam T Model parameter numeric type.
+ */
+private[bigdl] class TensorArraySize[T: ClassTag]()(implicit ev: TensorNumeric[T])
+  extends Operation[Table, Tensor[Int], T] {
+
+  override def updateOutput(input: Table): Tensor[Int] = {
+    val handle = input[Tensor[String]](1)
+    require(handle.isScalar, "Handle of a TensorArray must be a scalar")
+
+    val tensorArray = TensorArray(handle.value())
+
+    output = Tensor.scalar[Int](tensorArray.size)
+    output
+  }
+}
+
+/**
+ * Delete the TensorArray from the context.
+ *
+ * @tparam T Model parameter numeric type.
+ */
+private[bigdl] class TensorArrayClose[T: ClassTag]()(implicit ev: TensorNumeric[T])
```

Comment:
Is this operation guaranteed to be appear on every graph that TensorArray exists? Otherwise, the TensorArray created will be never released.

Reply:
I found that `tf.map_fn` used TensorArrays and does not have TensorArrayClose.

Reply:
fixed